frontend/catalyst/pennylane_extensions.py

```diff
@@ -261,31 +309,123 @@ def circuit(x):
+    """
+
+    def _check(x, hint):
+        if isinstance(x, list):
+            return x
+        elif isinstance(x, tuple):
+            return list(x)
+        raise ValueError(f"jvp '{hint}' argument must be a list or a tuple, not {type(x)}")
+
+    params = _check(params, "params")
+    tangents = _check(tangents, "tangents")
+    fn: Differentiable = _bless_differentiable(f)
+    grad_params = _check_grad_params(method, h, argnum)
+    jaxpr = _make_jaxpr_differentiable(fn, grad_params, *params)
+    return jprim.jvp_p.bind(*(params + tangents), jaxpr=jaxpr, fn=fn, grad_params=grad_params)
+
+
+def vjp(f, params, cotangents, *, method=None, h=None, argnum=None):
+    """A Vector-Jacobian product for PennyLane/Catalyst.
+
+    This function allows the Vector-Jacobian Product of a hybrid quantum-classical function to be
+    computed within the compiled program.
+
+    Args:
+        f(DifferentiableLike): Function-like object to calculate JVP for
+        params(List[Array]): List (or a tuble) of f's arguments specifying the point to calculate
+                             VJP at. A subset of these parameters are declared as
+                             differentiable by listing their indices in the ``argnum`` paramerer.
+        cotangents(List[Array]): List (or a tuple) of tangent values to use in JVP. The list size
+                                 and shapes must match the size and shape of ``f`` outputs.
+        method(str): Differentiation method to use, same as in ``grad``.
+        h (float): the step-size value for the finite-difference (``"fd"``) method
+        argnum (Union[int, List[int]]): the params' indices to differentiate.
+
+    Returns (Tuple[Array]):
+        Return values of ``f`` paired with the JVP values.
 
-    if isinstance(f, catalyst.compilation_pipelines.QJIT):
-        # Don't generate an extra function when the circuit is already qjitted.
-        f = f.qfunc
+    Raises:
+        TypeError: invalid parameter types
+        ValueError: invalid parameter values
+
+    **Example**
 
-    if isinstance(f, qml.QNode):
-        return Grad(f, method=method, h=h, argnum=argnum)
+    .. code-block:: python
+
+        def f(p):
+          return jnp.stack([1*p, 2*p, 3*p])
+
+        @qjit
+        def workflow(params, cotangent):
+            return vjp(f, [params], [cotangent])
+
+    >>> workflow(jnp.zeros([4], dtype=float), jnp.ones([3,4], dtype=float))
+    (Array([[0., 0., 0., 0.],
+            [0., 0., 0., 0.],
+            [0., 0., 0., 0.]], dtype=float64),
+     Array([6., 6., 6., 6.], dtype=float64))
+    """
 
-    return Grad(Function(f), method=method, h=h, argnum=argnum)
+    def _check(x, hint):
+        if isinstance(x, list):
+            return x
+        elif isinstance(x, tuple):
+            return list(x)
+        raise ValueError(f"vjp '{hint}' argument must be a list or a tuple, not {type(x)}")
+
+    params = _check(params, "params")
+    cotangents = _check(cotangents, "cotangents")
```

Comment:
Do you mean that we shouldn't distinguish iterables and non-iterables? The `_check` not only converts tuples to lists, but also checks that its parameter is a container of arguments (e.g. it raises on scalars). On top of that, the function saves us a few lines of template code.

Reply:
Well the function only checks whether the arguments are a list or a tuple, not whether they are a container in general. This type of hardcoded type checking is generally discouraged in Python, because as long the object passed in behaves like an iterable we don't care what type it actually is.

You *can* do an instance check against `Iterable` from the collections module, but even so I think it would be bloat and un-pythonic (it's just not necessary here given that Python will raise an error if necessary as soon as we unpack the arguments, and we don't need the arguments to be a list specifically).

This sort of check would make sense if we wanted to normalize/canonicalize an input that could be both a scalar and a list, wrapping the scalar into a list for easier treatment in subsequent code would make sense there.

Reply:
OK, then I choose to check against an iterable instance!